go/src/koding/klient/machine/index/change.go

```diff
@@ -0,0 +1,163 @@
+package index
+
+import (
+	"sync/atomic"
+	"time"
+)
+
+// ChangeMeta indicates what change has been done on a given file.
+type ChangeMeta uint64
+
+const (
+	ChangeMetaLocal  ChangeMeta = 1 << iota // local->remote synchronization.
+	ChangeMetaRemote                        // remote->local synchronization.
+	ChangeMetaAdd                           // File was added.
+	ChangeMetaRemove                        // File was removed.
+	ChangeMetaUpdate                        // File was updated.
+	ChangeMetaLarge                         // File size is above 4GB.
+)
+
+// Followed constants are helpers for ChangeMeta.Coalesce method.
+const (
+	cmInv = 0
+	cmEv  = ChangeMetaUpdate | ChangeMetaRemove | ChangeMetaAdd
+	cmAll = cmEv | ChangeMetaRemote | ChangeMetaLocal
+
+	cmUL = ChangeMetaUpdate | ChangeMetaLocal
+	cmDL = ChangeMetaRemove | ChangeMetaLocal
+	cmAL = ChangeMetaAdd | ChangeMetaLocal
+
+	cmUR = ChangeMetaUpdate | ChangeMetaRemote
+	cmDR = ChangeMetaRemove | ChangeMetaRemote
+	cmAR = ChangeMetaAdd | ChangeMetaRemote
+)
+
+// udarlMap is a helper array used to map coalesced changes to new change. It
+// has all((3+2)^5) combinations of UDA events and RL directions.
+var udarlMap = [32]ChangeMeta{
+	cmInv, cmInv, cmInv, cmInv,
+	cmAL, cmAL, cmAR, cmUL,
+	cmDL, cmDL, cmDR, cmDL,
+	cmUL, cmUL, cmUR, cmAll,
+	cmUL, cmUL, cmUR, cmUL,
+	cmAL, cmAL, cmAR, cmUL,
+	cmDL, cmDL, cmDR, cmAll,
+	cmInv, cmInv, cmInv, cmInv,
+}
+
+// Coalesce coalesces two meta-data changes and saves the result to called
+// object. The rules of coalescing are:
+//
+//  U - update; D - remove(delete); A - add; L - local; R - remote.
+//
+//      | UL | DL | AL | UR | DR | AR |
+//      +----+----+----+----+----+----+----
+//      | UL | DL | AL | UL | AL | UL | UL
+//      +----+----+----+----+----+----+----
+//           | DL | UL | DL | DL | DL | DL
+//           +----+----+----+----+----+----
+//                | AL | UL | AL | UL | AL
+//                +----+----+----+----+----
+//                     | UR | DR | AR | UR
+//                     +----+----+----+----
+//                          | DR | UR | DR
+//                          +----+----+----
+//                               | AR | AR
+//                               +----+----
+//
+// Example: If add remote change(AR) is merged with update local (UL) change,
+//          the coalesced event will be update local AR+UL=UL. This means that
+//          local updated file should overwrite remotely added one.
+//
+// All other flags are OR-ed. The coalesce matrix must be kept triangular.
+func (cm *ChangeMeta) Coalesce(newer ChangeMeta) {
+	for {
+		older := ChangeMeta(atomic.LoadUint64((*uint64)(cm)))
+
+		partial := udarlMap[(older|newer)&31]
```

Comment:
Instead of 31 you could:

```go
const all = ChangeMetaLocal | ChangeMetaRemote | ChangeMetaAdd | ChangeMetaRemove | ChangeMetaUpdate
```

and then `partial := udarlMap[(older|newer)&all]` here.